### ISSUE TITLE
[Homekit] Proposal to extend support for installation with >150 Accessories

### ISSUE THREAD

**Issue by kgoderis** [openhab/openhab-addons#12927]
The HomeKit Accessory Protocol Specification (Non-Commercial Version Release R2) specifies that the maximum number of Accessories a Bridge can return is 150. 

Due to the fact that org.openhab.io.homekit is linking into HAP at the Item level (instead of Thing level) very rapidly this limit is hit in installation with a lot of Items. Hence the need to find a solution to overcome this limit. 

I am offering some effort to implement such a solution but want to discuss this upfront before spending any time on it. I happen to have a (private) custom implementation of the Homekit bundle that not only exposes openHAB as an HAP Accessory Server, but also as a HAP Client, allowing openHAB to control Homekit devices. Furthermore, it ties into HAP at the Thing level (e.g. Things ma on HAP Accessories, Channels map on HAP Characteristics). Anyways, that being said, my proposal is as follows:

1. Extend the bundle so that it supports more than one HAP Bridge. Each Bridge (see later) is running on a separate TCP port. In the existing configuration, the port parameter is used as the first port number to use for the first Bridge, all other Bridges are using the next available port number, e.g. port+1, port+2,...

2. Define an optional configuration parameter to designate an Item to use a specific Bridge.

```
Rollershutter window_covering "Window Rollershutter" {homekit = "WindowCovering"  [bridge="myFirstBridge"]}
Rollershutter another_window_covering "Window Rollershutter" {homekit = "WindowCovering"  [bridge="myOtherBridge"]}
```

If no Bridge is specified, then it is considered to be the "Default" Bridge

As such, it is up to the User to make sure that less than 150 items are assigned to a Bridge. And it allows to use to make any kind of combination of Accessories. For example, group all Outlets in a Bridge, or group all Accessories of a given floor level into a Bridge, and so forth

3. The Homekit ItemRegistryChangeListener implementation is responsible for managing these Bridges

4. Update the CommandExtension and GUI to reflect a multiple-server setup, e.g. facilitate more than 1 QR code for pairing and so forth

**Comment by yfre** [comment]
@kgoderis 
this is the approach i was thinking of as well - multiple bridges with tagging at accessories to which bridges the belong. the Java-HAP libraries supports already multiple instances  (with different ports, different mDNS records, etc). 
this would not only help with larger installation but make home app faster / more responsive as one could have multiple bridges with just 5-10 accessories and home app can update their status in parallel. 

but i never found a time to implement this.  so, feel to start and i can contribute. 


to your other points:
- having homekit client would be great, but it is imo significant effort
- not sure i understood your comment regarding items and things.  currently we have
 thing -> channel -> item -> homekit characteristics  and then items group -> homekit accessory. so, e.g. hue lights bundles multiple items to one homekit accessory.

**Comment by jlaur** [comment]
> I happen to have a (private) custom implementation of the Homekit bundle that not only exposes openHAB as an HAP Accessory Server, but also as a HAP Client, allowing openHAB to control Homekit devices.

I'm not into HomeKit, but if I understand this correctly, this would mean that devices with HomeKit support would be able to integrate with openHAB by simulating a HomeKit Controller? Perhaps something like this: https://github.com/jlusiardi/homekit_python?

If this is true, it would mean that I could integrate my Velux KIG 300 because it supports HomeKit?

If you have something working, or when you would have something working, I would be very interested in testing this. :-)

**Comment by yfre** [comment]
> > I happen to have a (private) custom implementation of the Homekit bundle that not only exposes openHAB as an HAP Accessory Server, but also as a HAP Client, allowing openHAB to control Homekit devices.
> 
> 
> 
> I'm not into HomeKit, but if I understand this correctly, this would mean that devices with HomeKit support would be able to integrate with openHAB by simulating a HomeKit Controller? Perhaps something like this: https://github.com/jlusiardi/homekit_python?
> 
> 
> 
> If this is true, it would mean that I could integrate my Velux KIG 300 because it supports HomeKit?

Exactly, with HomeKit client you could integrate with openHAB all devices that support HomeKit

**Comment by kgoderis** [comment]
@yfre Indeed, building a Homekit client is a considerable effort, but I have done exactly that. However, in the same effort I rewrote the complete current org.openhab.io.homekit bundle so that it would both be a client and a server. In order to release a client that fits with the existing bundle, I would have to refactor a lot of stuff. 

In essence, I also rewrote HAP-java to remove the API's that sit between openHAB and HAP-java. 

IMO, "routing" the bridge between HAP and openHAB through Items is not ideal, and therefore I wrote all the functionality that automatically (!) exposes HAP Accessories based on Things being provided by the configuration files and alike.

**Comment by yfre** [comment]
sound good. maybe you can submit it as alternative homekit implementation and  over time users would migrate to it. 
however, personally i would have an issue with exposing things as homekit accessories:
- i dont want to expose all things to homekit. in addition, many of them are not supported by homekit, e.g. dishwasher, weather station, car
- i have often homekit accessories which have link to multiple things, e.g. one accessory like heating combines 2 devices/things. or in knx one thing is always very simple and only one characteristic. 
- some of the characteristics are virtual, i.e. not connected to any things  / real hardware, but used to trigger rules
- i also have complex things which are exposing multiple homekit accessories, e.g. some channels are one accessory, another channels are second

here is an example of my heating (it connects to 2 KNX things and to 2 virtual). you are also see alexa follows the same approach.
```
Group gThermostat {homekit="Thermostat", alexa="Thermostat"}
Number currentTemp "Current temperature [%.1f C]"            (gThermostat) {homekit="CurrentTemperature", alexa="CurrentTemperature", channel="knx:device:scn_ip:GT2T_livingroom:temperature"}
Number targetTemperature "Target Temperature [%.1f C]" (gThermostat) {homekit="TargetTemperature", alexa="TargetTemperature", channel="knx:device:scn_ip:AKH800_01:channel_a_solltemp"}
String currentMode "Current Mode"            (gThermostat) {homekit="CurrentHeatingCoolingMode" [OFF="OFF", HEAT="HEAT"], alexa="HeatingCoolingMode" [OFF="OFF",HEAT="HEAT"]}
String targetMode "Target Mode"        (gThermostat) {homekit="TargetHeatingCoolingMode" [OFF="OFF", HEAT="HEAT"]}
```

in general, my understanding of openhab concept is - users interact with items (e.g. via main ui or alex or google or homekit) and items send command to thing, which are representing actually device. this gives the flexibility of combining multiple things to one item and other way around, having multiple items from one thing.

personally, i would fully support a new implementation of homekit as the current one has some design issues, but only if the concept is as flexible as current and allows to support all different scenarios.

**Comment by kgoderis** [comment]
> * i dont want to expose all things to homekit. in addition, many of them are not supported by homekit, e.g. dishwasher, weather station, car

Ack. In fact, what I wrote works by mapping homekit devices to things through java configuration classes that can be maintained. Not everything is exposed, but it is done in a more dynamic way, e.g. not through a config file

> * i have often homekit accessories which have link to multiple things, e.g. one accessory like heating combines 2 devices/things. or in knx one thing is always very simple and only one characteristic.
> * some of the characteristics are virtual, i.e. not connected to any things  / real hardware, but used to trigger rules
> * i also have complex things which are exposing multiple homekit accessories, e.g. some channels are one accessory, another channels are second

I agree with should not do away with the current functionality/flexibility.

> personally, i would fully support a new implementation of homekit as the current one has some design issues, but only if the concept is as flexible as current and allows to support all different scenarios.

Can you highlight those design issues? Maybe there is solution for them

**Comment by kgoderis** [comment]
@wborn @kaikreuzer @J-N-K Question to you guys. Cfr all the above, building a Homekit Client bundle for openHAB, as opposed to being only a Homekit Server, implies giving up the dependency on HAP-java as a lot of the code at the client and server side is very similar (for example, the whole encrypted pairing setup/protocol). Are you/we ok that we depart from HAP-java and have our own home-grown implementation of HAP?

**Comment by yfre** [comment]
> Can you highlight those design issues? Maybe there is solution for them

the main one:  missing accessory information in HomekitCharacteristicFactory

the binding only links the Java-HAP characteristics to openHab item callbacks, e.g. 
TargetFanStateCharacteristic(openHabItem.getState(), openHabItem.setState()). 
that is all what i do for fan state.  one line. 

JavaHAP gets HAP request for accessories and calls the callback directly, done. that makes is very light, scalable and very performant. Only few lines of code are executed per characteristic. 

but there are few (3-4) accessories which have dependencies between characteristics, e.g. battery charging state acts differently for chargeable and non-chargeable batteries.  in such case  the life of developer would easier if the method for characteristic creation would not only get the single openHAB as input but the complete group. 

not sure whether i explain it well. but is mainly about missing parameters in methods signatures.

**Comment by ccutrer** [comment]
Given the title of this issue, and the contents, I think it should be broken up into multiple issues.

#13226 will close this, as the primary topic of the issue - breaking the 150 limit.

For OpenHAB controlling native HomeKit accessories should be its own issue, and I would propose that it should be a separate binding bundle, and not necessarily part of the HomeKit IO addon (though it's fine if they have an interdependence or a common library, similar to how the MQTT binding is actually made up of several bundles).

As for tying HomeKit accessories more directly to things, I disagree. The OpenHAB model is definitely to decouple the model of what an OpenHAB user interacts with from the underlying connections to physical devices. While it seems like a lot of duplication at first, it's very powerful to be able to shape things and not have to re-write rules, sitemaps, etc. as the underlying implementation evolves. For example, changing which binding represents a device, abstracting to or from a single device to a group of devices, changing which device provides a piece of information, aggregating multiple physical devices into a single HomeKit accessory, disaggregating a single physical device into multiple HomeKit accessories, etc. etc. (several of which @yfre already mentioned). That said, I don't think it's a bad idea to make it easier for the "simple" use cases by having a dropdown or something on a thing, that for example is a zwave switch, and you just choose "Light", and it automatically adds the semantic tags (for the semantic model), auto-creates items for the first dimmer or switch channel it finds, and adds HomeKit (and Alexa and google home) metadata linking to light accessory (and brightness or color characteristics, etc. as necessary). But that would be separate from the HomeKit server implementation itself.
